### PR TITLE
[BUILDER] Simple builder

### LINKED ISSUE #2559: [BUILDER] - Simple builder
### What is this task and why do we need to work on it?

For builder integration we need a simple builder internal to HotShot that fetches transactions from CDN and uses those to build blocks.

### What work will need to be done to complete this task?

_No response_

### Are there any other details to include?

_No response_

### What are the acceptance criteria to close this issue?

Simple builder implemented & tested against builder client
Includes tests to confirm builder correctly fetches and includes transactions

### Branch work will be merged to (if not the default branch)

_No response_

### PR DESCRIPTION
**Pull request by QuentinI**
Closes #2559 

### This PR: 
Adds a simple builder for tests that aims to build blocks the same way HotShot does it now.

### This PR does not: 
Integrate said builder.

### CODE REVIEW

File: crates/testing/src/block_builder.rs [hunk 1]
```diff
@@ -1,36 +1,151 @@
...
+    }
+
+    /// Spawn a task building blocks, configured with given options
+    #[allow(clippy::missing_panics_doc)] // ony panics on 16-bit platforms
+    pub fn run(&self, options: RandomBuilderOptions) {
+        let blocks = self.blocks.clone();
+        let (priv_key, pub_key) = (self.priv_key.clone(), self.pub_key);
+        async_spawn(async move {
+            let mut rng = SmallRng::from_entropy();
+            let time_per_block = Duration::from_millis(1000 / options.blocks_per_second);
```
Comment (jparr721) [inline crates/testing/src/block_builder.rs:108]: The random number here is a bit hard to parse. Could it perhaps be a named constant? For example, is 1000 the block epoch time? It's hard to know in this case.

**Review by jparr721 — COMMENTED** [review]
A few cosmetics

File: crates/testing/src/block_builder.rs (hunk 1 above)
Comment (QuentinI) [reply crates/testing/src/block_builder.rs:108]: Right, I already have no idea what this is about either :)

**Comment by QuentinI** [comment]
Sorry for force-pushing, couldn't figure out how to update this branch otherwise, since it included a commit already merged to main now.

File: crates/testing/src/block_builder.rs (hunk 1 above)
Comment (QuentinI) [reply crates/testing/src/block_builder.rs]: I re-wrote it as `Duration::from_secs(1) / options.blocks_per_second`, that should make it clear hopefully.

File: crates/testing/src/block_builder.rs [hunk 2]
```diff
@@ -279,3 +225,244 @@ pub fn run_random_builder(url: Url) {
...
+            }
+            HotShotEvent::LeafDecided(leaf_chain) => {
+                let mut queue = this.transactions.write().await;
+                for leaf in leaf_chain.iter() {
+                    if let Some(ref payload) = leaf.block_payload {
+                        for txn in payload.transaction_commitments(&()) {
+                            queue.remove(&txn);
+                        }
+                    }
+                }
```
Comment (bfish713) [inline crates/testing/src/block_builder.rs:356]: This logic won't remove transactions if the `TransactionRecv` event happens after a leaf containing that transaction is decided.  It's pretty common with the webserver we download transactions in batches right before we become a leader.  So we'll get lots of transactions that were already decided.  I'm not sure how relevant that is for this builder if we only use as a test.  If we do bench-marking with it then we probably need to handle this case.

**Review by bfish713 — APPROVED** [review]
Looks good, left one comment we might need to fix.  Can be a separate PR
